src/Microsoft.AspNetCore.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -377,6 +386,36 @@ protected virtual string GenerateUniqueId()
             }
         }
 
+        private void RenderPreallocatedAttributeValues(
+            PreallocatedTagHelperAttributeChunk preallocatedAttributeValueChunk,
+            string valueAccessor)
+        {
+            var attributeValueAccessor = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0}.{1}",
+                    preallocatedAttributeValueChunk.AttributeVariableAccessor,
+                    _tagHelperContext.TagHelperAttributeValuePropertyName);
+
+            _writer
+                .WriteStartAssignment(valueAccessor)
+                .WriteInstanceMethodInvocation(
+                    attributeValueAccessor,
+                    nameof(attributeValueAccessor.ToString));
```

Comment:
I was just worried that this would allocate, but it doesn't as you clarified. I kind of like the casting approach now that I look at it. Ensures that the system is working end-to-end.